File: los-service/src/main/java/com/mal/lospoc/los/api/LoanController.java
package com.mal.lospoc.los.api;

import com.mal.lospoc.common.domain.LoanApplication;
import com.mal.lospoc.common.dto.UserDetails;
import com.mal.lospoc.los.application.service.LoanApplicationService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api")
public class LoanController {
    private final LoanApplicationService service;

    public LoanController(LoanApplicationService service) {
        this.service = service;
    }

    @PostMapping("/applications")
    public ResponseEntity<Map<String, Object>> submit(@RequestBody SubmitRequest req) {
        UUID appId = service.submit(req.productId(), req.userDetails(), req.loanAmount());
        return ResponseEntity.ok(Map.of("applicationId", appId.toString(), "status", "submitted"));
    }

    @GetMapping("/applications/{id}")
    public ResponseEntity<LoanApplication> get(@PathVariable UUID id) {
        return ResponseEntity.ok(service.get(id));
    }

    record SubmitRequest(String productId, UserDetails userDetails, BigDecimal loanAmount) {}
}

File: src/main/kotlin/org/example/controller/LoanController.kt
package org.example.controller

import dev.restate.client.Client
import org.example.model.LoanApplication
import org.example.model.LoanResult
import org.example.workflow.LoanApplicationWorkflowClient
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal
import java.util.UUID

data class CheckCreditRequest(
    val applicantName: String,
    val amount: BigDecimal,
    val income: BigDecimal
)

@RestController
@RequestMapping("/api")
class LoanController(
    @Value("\${restate.endpoint.url:http://localhost:8080}") private val restateUrl: String
) {
    private val logger = LoggerFactory.getLogger(LoanController::class.java)

    // Lazy initialization of Restate client
    private val restateClient: Client by lazy {
        Client.connect(restateUrl)
    }

    @PostMapping("/checkCredit")
    suspend fun checkCredit(@RequestBody request: CheckCreditRequest): LoanResult {
        val applicationId = UUID.randomUUID().toString().take(8)

        val application = LoanApplication(
            applicationId = applicationId,
            applicantName = request.applicantName,
            amount = request.amount,
            income = request.income
        )

        logger.info("Processing loan application via Restate: {}", application.applicationId)

        // Call workflow via Restate client
        val workflowClient = LoanApplicationWorkflowClient.fromClient(restateClient, applicationId)
        val result = workflowClient.processApplication(application)

        logger.info("Loan application {} completed with decision: {}", applicationId, result.decision)

        return result
    }
}
